embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,887 @@
+    /// with [`send`], this function has two failure cases instead of one (one for
+    /// disconnection, one for a full buffer).
+    ///
+    /// # Errors
+    ///
+    /// If the channel capacity has been reached, i.e., the channel has `n`
+    /// buffered values where `n` is the argument passed to [`channel`], then an
+    /// error is returned.
+    ///
+    /// If the receive half of the channel is closed, either due to [`close`]
+    /// being called or the [`Receiver`] handle dropping, the function returns
+    /// an error. The error includes the value passed to `send`.
+    ///
+    /// [`send`]: Sender::send
+    /// [`channel`]: channel
+    /// [`close`]: Receiver::close
+    pub fn try_send(&self, message: T) -> Result<(), TrySendError<T>> {
+        self.channel.get().try_send(message)
+    }
+
+    /// Completes when the receiver has dropped.
+    ///
+    /// This allows the producers to get notified when interest in the produced
+    /// values is canceled and immediately stop doing work.
+    pub async fn closed(&self) {
+        CloseFuture {
+            sender: self.clone(),
+        }
+        .await
+    }
+
+    /// Checks if the channel has been closed. This happens when the
+    /// [`Receiver`] is dropped, or when the [`Receiver::close`] method is
+    /// called.
+    ///
+    /// [`Receiver`]: crate::sync::mpsc::Receiver
+    /// [`Receiver::close`]: crate::sync::mpsc::Receiver::close
+    pub fn is_closed(&self) -> bool {
+        self.channel.get().is_closed()
+    }
+}
+
+struct SendFuture<'ch, M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    sender: Sender<'ch, M, T, N>,
+    message: UnsafeCell<T>,
+}
+
+impl<'ch, M, T, const N: usize> Future for SendFuture<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    type Output = Result<(), SendError<T>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self
+            .sender
+            .channel
+            .get()
+            .try_send_with_context(unsafe { self.message.get().read() }, Some(cx))
+        {
+            Ok(..) => Poll::Ready(Ok(())),
+            Err(TrySendError::Closed(m)) => Poll::Ready(Err(SendError(m))),
+            Err(TrySendError::Full(..)) => {
+                Poll::Pending
+                // Note we leave the existing UnsafeCell contents - they still
+                // contain the original message. We could create another UnsafeCell
+                // with the message of Full, but there's no real need.
+            }
+        }
+    }
+}
+
+struct CloseFuture<'ch, M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    sender: Sender<'ch, M, T, N>,
+}
+
+impl<'ch, M, T, const N: usize> Future for CloseFuture<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.sender.channel.get().is_closed_with_context(Some(cx)) {
```

Comment:
Why does `close` have to block until close is "done"? In other channels I've seen, `close` completes instantly, even if the channel still has queued items. Receivers still receive these items, and receive the "closed" error when there are no more queued items.

It seems more intuitive to me: `send` doesn't wait for receivers to receive the item, so `close` shouldn't wait for receivers to receive the close.

Reply:
Note that the receiver's `close` method completes immediately, whereas the sender's `closed` method completes when the receiver is dropped. Its Tokio API doc states, "This allows the producers to get notified when interest in the produced values is canceled and immediately stop doing work.".